File: 216. Combination Sum III.cpp
class Solution {
private:
    vector<vector<int>> res;
public:
    vector<vector<int>> combinationSum3(int k, int n) {
        DFS({}, 1, k, n);
        return res;
    }
    void DFS(vector <int> path, int start, int k, int n)
    {
        if (path.size() == k || n == 0 || start == 10)
        {
            if (path.size() == k && n == 0)
            {
                res.push_back(path);
                return;
            }
        }
        for (int i = start; i <= 9; i++)
        {
            if (i <= n)
            {
                path.push_back(i);
                DFS(path, i + 1, k, n - i);
                path.pop_back();
            }
        }
    }
};

File: 216. Combination Sum III.py
class Solution(object):
    def combinationSum3(self, k, n):
        """
        :type k: int
        :type n: int
        :rtype: List[List[int]]
        """
        self.res = []
        self.DFS([], 1, k, n)
        return self.res
        
    def DFS(self, path, start, k, n):
        if len(path) == k or n == 0 or start == 10:
            if len(path) == k and n == 0:
                self.res += [path[:]]
            return
        for i in xrange(start, 10):
            if i <= n:
                self.DFS(path + [i], i+1, k, n-i)
